src/render/websocket.js

```diff
@@ -0,0 +1,190 @@
+const headers = require('./headers')
+const cookies = require('./cookies')
+const address = require('./address')
+const object = require('./object')
+const { websocketFactor: makeWebsocketFactor } = require('../make')
+
+function websocket(spec) {
+  let factor = getFactor(spec.request, spec)
+  return render(factor)
+}
+
+function getFactor(request, spec) {
+  const factor = makeWebsocketFactor()
+  if (spec.addSleep) {
+    factor.timeAlive = timeAlive(spec.timeConnected, spec.webSocketMessages)
+    factor.addSleep = spec.addSleep
+    factor.messages = changeMessageTimes(spec.webSocketMessages)
+  } else {
+    factor.messages = spec.webSocketMessages
+  }
+  factor.call = 'connect'
+  address(request, factor)
+  factor.headers = headers(request.headers)
+  factor.cookies = cookies(request.cookies)
+  factor.options = options(factor)
+  factor.args = args(factor)
+
+  return factor
+}
+
+function changeMessageTimes(messages) {
+  /// calculates when to sleep between messages
+  if (messages.length != 0) {
+    let firstTimeStartsAtZero = 0
+    let firstIndexSetToZero = 0
+    let messagesWithNewTimes = JSON.parse(JSON.stringify(messages))
+    messagesWithNewTimes[firstIndexSetToZero].time = firstTimeStartsAtZero
+    for (let i = 1; i < messages.length; i++) {
+      messagesWithNewTimes[i].time =
+        (messages[i].time - messages[i - 1].time) * 1000 // 1000 millisecond multiplier
+    }
+    var sum = messagesWithNewTimes.reduce((accum, curr) => {
+      return accum + curr.time
+    }, 0)
+    for (var i = messagesWithNewTimes.length - 1; i >= 0; i--) {
+      sum = sum - messagesWithNewTimes[i].time
+      messagesWithNewTimes[i].time = sum
+    }
+    return messagesWithNewTimes
+  }
+}
+
+function timeAlive(timeConnected, messages) {
+  /// calculates length of time when websocket is open minus the total time the messages took to run before closing the socket
+  let time = timeConnected
+  if (messages.length > 1) {
+    let firstMessageTime = messages[0].time
+    let finalMessageTime = messages[messages.length - 1].time
+    let totalTimeMessagesTook = finalMessageTime - firstMessageTime
+    time = timeConnected - totalTimeMessagesTook
+  }
+  return time
+}
+
+function ws_send_messages(factor) {
```

Comment:
This function appears to be pretty GraphQL/PassiveLogic specific (referencing 44XX response codes as errors, using connection_init/connection_ack messages, etc). Is that the intent, or are these changes intended to be generic for all websocket traffic?

Reply:
ah, yes. You are correct. Will fix.